

#activate packages
```{r}
library(dada2)
library(phyloseq)
library(Biostrings)
library(ggplot2)
library(tibble)
```



#set dir
```{r}
path <- "/path/to/data" # CHANGE ME to the directory containing the fastq files after unzipping.
list.files(path)
```

#create lists of forvard and reverse reads
```{r}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_R"), `[`, 1)


```

#Filter and trim
```{r}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```

```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)

#Merge paired reads
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
#head(mergers[[1]])

#Construct sequence table
seqtab <- makeSequenceTable(mergers)
dim(seqtab)

# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))

#Remove chimeras
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

#Track reads through the pipeline
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

#Assign taxonomy
```{r}
taxa <- assignTaxonomy(seqtab.nochim, "/path/to/GTDB_bac120_arc122_ssu_r202_Genus.fa.gz", multithread=TRUE)
```

#save results
```{r}
df_seq_no_chim <- as.data.frame(seqtab.nochim)
df_taxa <- as.data.frame(taxa)

df_taxa <- df_taxa %>% filter(Kingdom %in% c("Bacteria"))
df_seq_no_chim <- df_seq_no_chim %>% select(rownames(df_taxa))

track <- as.data.frame(track)
track <- rownames_to_column(track)
track$sampleid <- track$rowname
track <- track[, c(8, 2:7)]

write.table(df_seq_no_chim, "/path/to/seqtab.nochim.tsv", quote = F, sep = "\t")
write.table(df_taxa, "/path/tol/taxa.tsv", quote = F, sep = "\t")
write.table(track, "/path/to/statistics.tsv", quote = F, sep = "\t")
```


